### DISCUSSION TITLE
Can MLIR APIs be more Pythonic and extensible?

### DISCUSSION THREAD

**Discussion by zhwa** [modular/modular#154] [Ideas]
Thanks the Mojo team for this great new language. It's really amazing!

Reading the documents on MLIR related APIs, I feel that the style of these APIs seems to be quite different with Python. For example, in the following official example, syntaxes like ```Self {}```, ``` `index.bool.constant` ``` and ```[value : --mlir_attr.`false`,]()``` can rarely be seen in Python code.

```py
@register_passable("trivial")
struct OurBool:
    var value: __mlir_type.i1

    fn __init__() -> Self:
        return Self {
            value: __mlir_op.`index.bool.constant`[
                value : __mlir_attr.`false`,
            ]()
        }
```

So, I wander, can the API be more Pythonic? Like:

```py
import MLIR

@register_passable("trivial")
struct OurBool:
    var value: MLIR.type.i1

    fn __init__() -> Self:
        return Self {  MLIR.index.bool(MLIR.type.false)  }
```

Also, the doc mentioned that in the future, Mojo will support more dialects from MLIR.  I wander if it is possible to make Mojo more extensible such that it can also create new didacts?

Thanks!

**Comment by lattner** [comment]
Indeed, the MLIR integration hasn't been polished or designed to be pretty - we've focused primarily on making it fully capable and unblocking our needs.  The idea for it is that only MLIR experts would be using this, but then they'd be wrapping user-facing Pythonic types and methods around them (e.g. like OurBool wraps i1).  that said, we can definitely improve this in various ways, we just can't do so at the loss of fidelity/expressiveness.

**Comment by lattner** [comment]
>  I wonder if it is possible to make Mojo more extensible such that it can also create new didacts?

This is also something we're likely to look into in the far future, but isn't a priority right now.  Also, as mojo opens up more, it would be great for community members to poke at this.

**Reply by strangemonad** [reply]
+1 to this. I'd be curious to see how well mojo could behave as an alternative to TableGen (or be lowered to TableGen) to define dialects.

**Reply by lattner** [reply]
Given that tablegen is my fault, I think it's ok for me to say that I would love to see tblgen die :-) :-)

**Reply by strangemonad** [reply]
😂 I mean if a large amount of value under the fault's lifetime usage curve was delivered until something better comes along then 🤷‍♂️

**Reply by lattner** [reply]
fair enough!
